### PR TITLE
CDev::poll() performance improvements

### PR DESCRIPTION
**Pull request by bkueng**
When looking at the uorb latency test with `uorb_tests latency_test` the latency for publish-poll-wakeup is 40 us. This seems small, but running at 1kHz it accounts for 4% CPU load for a single pub-sub pair.

This PR reduces the latency to 30us by avoiding the use of an expensive semaphore and disabling IRQs instead (the actual critical section is very small).
It also improves the `uorb_tests latency_test` for better output and more consistent results.

master:
```
uORB note: ---------------- LATENCY TEST ------------------
INFO  [uorb_tests] mean:     40.4320 us
INFO  [uorb_tests] std dev:   1.3466 us
INFO  [uorb_tests] min:      39 us
INFO  [uorb_tests] max:      57 us
INFO  [uorb_tests] missed topic updates: 0
```
This PR:
```
uORB note: ---------------- LATENCY TEST ------------------
INFO  [uorb_tests] mean:     30.4120 us
INFO  [uorb_tests] std dev:   1.0637 us
INFO  [uorb_tests] min:      29 us
INFO  [uorb_tests] max:      45 us
INFO  [uorb_tests] missed topic updates: 0
```

The changes here will become relevant when increasing the rate controller loop to 1kHz. I quickly tested running the attitude controller and fmu at 1kHz and the difference in CPU load is 4% (35 vs 31).

I've been using and flying with these changes for over a week.

Changes are split into separate commits with further explanations.

### Jitter

One interesting side note is looking at the latency after stopping mavlink (`mavlink stop-all`):
```
uORB note: ---------------- LATENCY TEST ------------------
INFO  [uorb_tests] mean:     30.1050 us
INFO  [uorb_tests] std dev:   0.3288 us
INFO  [uorb_tests] min:      29 us
INFO  [uorb_tests] max:      34 us
INFO  [uorb_tests] missed topic updates: 0
```
It can be seen that there is quite a bit less jitter that comes from UART interrupts/critical sections. I think having UART TX DMA would help here.

### CODE REVIEW

File: src/lib/drivers/device/CDev.cpp [hunk 1]
```diff
@@ -338,31 +396,22 @@ CDev::poll_notify(pollevent_t events)
 void
 CDev::poll_notify_one(px4_pollfd_struct_t *fds, pollevent_t events)
 {
-	DEVICE_DEBUG("CDev::poll_notify_one");
-
-#ifdef __PX4_NUTTX
-	int value = fds->sem->semcount;
```
Comment (dagar) [inline src/lib/drivers/device/CDev.cpp]: Any idea why this was here in the first place?

**Comment by dagar** [comment]
Looks good.

Just a thought, but if we really wanted to optimize orb polling I think we could get away with a single semaphore per cdev and avoid the bulk of poll waiter setup and teardown (at least on nuttx).

**Comment by dagar** [comment]
> It can be seen that there is quite a bit less jitter that comes from UART interrupts/critical sections. I think having UART TX DMA would help here.

@davids5 FYI

File: src/lib/drivers/device/CDev.cpp (hunk 1 above)
Comment (bkueng) [reply src/lib/drivers/device/CDev.cpp]: My guess is that the original author thought it's not good to unlock the semaphore twice.

**Comment by davids5** [comment]
@bkueng 
If this is just a uOrb - should it be moved there and done differently? 
Is the dynamic allocation warranted?
px4_enter/exit_critical section is a macros and does not need #ifdeffing.
the disable and re enable around the malloc may not be needed. What are we trying to do? Avoid a context switch? Would schedule lock be a better choice?

File: src/lib/drivers/device/CDev.cpp (hunk 1 above)
Comment (bkueng) [reply src/lib/drivers/device/CDev.cpp]: > The next waiter did not wait when it should have. This should be investigated.

There is no next waiter, the semaphore is only used for that specific poll() call. But yes, please go through the code.

**Comment by bkueng** [comment]
> @bkueng out of curiosity did you fly at 1kHz? If so could you share the log?

I obviously had to test it :) But you don't want to see the code. It just the bare minimum required for acro and in-flight switching between 250 and 1000 Hz. A log is here: https://logs.px4.io/plot_app?log=74d50cdc-3ee6-4534-bbfe-2e9576514323
As for the flight behavior: to my surprise it does make a small difference. You'll notice it in situations where you have strong prop-wash effects for instance (the vehicle is more stable). I only flew LOS, I can imagine that the effect is stronger noticeable in FPV.
My goal is to implement it properly anytime soon, and after that we'll need the >=4kHz IMU sampling.

----

@davids5 The goal here is to maximize performance and minimize memory overhead. Which is why there's a dynamic allocation (that is rarely used). I was thinking about a linked list but there is nothing where we could add the next pointer (without changing NuttX).
malloc uses a semaphore, which is why the IRQs are re-enabled for that.

File: src/lib/drivers/device/CDev.cpp (hunk 1 above)
Comment (bkueng) [reply src/lib/drivers/device/CDev.cpp]: Just to be double sure I tested that:
- the semaphore counter goes up to 2 at this location in some cases
- at the same time no `poll()` does ever return early without an event

So it's not needed.

**Pull request merged**